src/main/java/org/reaktivity/nukleus/mqtt/internal/stream/MqttServerFactory.java

```diff
@@ -265,12 +300,375 @@ private RouteFW resolveTarget(
+            progress = publish.limit();
+        }
+
+        server.decoder = decodePacketType;
+        return progress;
+    }
+
+    private int decodeSubscribe(
+        MqttServer server,
+        final long traceId,
+        final long authorization,
+        final long budgetId,
+        final DirectBuffer buffer,
+        final int offset,
+        final int limit)
+    {
+        final MqttSubscribeFW subscribe = mqttSubscribeRO.tryWrap(buffer, offset, limit);
+        int progress = limit;
+        if (limit == 0 || subscribe == null)
+        {
+            server.doEncodeDisconnect(authorization, 0x82);
+            server.doNetworkEnd(traceId);
+        }
+        else
+        {
+            server.onDecodeSubscribe(authorization, subscribe);
+            progress = subscribe.limit();
+        }
+
+        server.decoder = decodePacketType;
+        return progress;
+    }
+
+    private int decodeUnsubscribe(
+        MqttServer server,
+        final long traceId,
+        final long authorization,
+        final long budgetId,
+        final DirectBuffer buffer,
+        final int offset,
+        final int limit)
+    {
+        final MqttUnsubscribeFW unsubscribe = mqttUnsubscribeRO.tryWrap(buffer, offset, limit);
+        int progress = limit;
+        if (limit == 0 || unsubscribe == null)
+        {
+            server.doEncodeDisconnect(authorization, 0x82);
+            server.doNetworkEnd(traceId);
+        }
+        else
+        {
+            server.onDecodeUnsubscribe(authorization, unsubscribe);
+            progress = unsubscribe.limit();
+        }
+
+        server.decoder = decodePacketType;
+        return progress;
+    }
+
+    private int decodePingreq(
+        MqttServer server,
+        final long traceId,
+        final long authorization,
+        final long budgetId,
+        final DirectBuffer buffer,
+        final int offset,
+        final int limit)
+    {
+        final MqttPingReqFW ping = mqttPingReqRO.tryWrap(buffer, offset, limit);
+        server.onDecodePingReq(authorization, ping);
+        server.decoder = decodePacketType;
+        return ping.limit();
+    }
+
+    private int decodeDisconnect(
+        MqttServer server,
+        final long traceId,
+        final long authorization,
+        final long budgetId,
+        final DirectBuffer buffer,
+        final int offset,
+        final int limit)
+    {
+        final MqttDisconnectFW disconnect = mqttDisconnectRO.tryWrap(buffer, offset, limit);
+        int progress = limit;
+        if (limit == 0 || disconnect == null)
+        {
+            server.doEncodeDisconnect(authorization, 0x82);
+            server.doNetworkEnd(traceId);
+        }
+        else
+        {
+            server.onDecodeDisconnect(traceId, disconnect);
+            progress = disconnect.limit();
+        }
+
+        server.decoder = decodePacketType;
+        return progress;
+    }
+
+    private int decodeIgnoreOne(
```

Comment:
Does the MQTT specification have explicit language indicating that an unrecognized packet type should be skipped and processing continue on subsequent packets?

If not, then we should remove the `decodeIgnoreOne` method.

Reply:
Self reminder:
```
in http2 we have onDecodeError trigger the cleanup for any existing application streams and also write out the GOAWAY frame, which is similar in spirit to mqtt DISCONNECT (edited) 
then after calling onDecodeError we set the decoder to decodeIgnoreAll
see doEncodeGoAway which for mqtt would be doEncodeDisconnect
new messages
then in doNetworkData after encode buffer is cleaned up, we check to see if no application streams are in flight and decoder is ignore-all, then we trigger doNetworkEnd
also in onNetworkEnd if decodeSlot is empty (so all pending frames have been processed then we cleanup any application streams in flight and doNetworkEnd
note: there is a TODO there to let the inbound complete normally and then let the outbound close on flush, but we don’t have that in place just yet and right now forcibly cleanup the application streams
this is typically a non-issue in practice, but we would prefer to make it better going forward, so recommend you keep the comment too as a reminder
```